parachain/src/wasm_executor/workspace.rs

```diff
@@ -0,0 +1,523 @@
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::primitives::{ValidationParams, ValidationResult};
+use parity_scale_codec::{Decode, Encode};
+use raw_sync::{
+	events::{Event, EventImpl, EventInit, EventState},
+	Timeout,
+};
+use shared_memory::{Shmem, ShmemConf};
+use std::{
+	error::Error,
+	fmt,
+	io::{Cursor, Write},
+	slice,
+	sync::atomic::AtomicBool,
+	time::Duration,
+};
+
+// maximum memory in bytes
+const MAX_PARAMS_MEM: usize = 1024 * 1024; // 1 MiB
+const MAX_CODE_MEM: usize = 16 * 1024 * 1024; // 16 MiB
+const MAX_VALIDATION_RESULT_HEADER_MEM: usize = MAX_CODE_MEM + 1024; // 16.001 MiB
+
+/// Params header in shared memory. All offsets should be aligned to WASM page size.
+#[derive(Encode, Decode, Debug)]
+struct ValidationHeader {
+	code_size: u64,
+	params_size: u64,
+}
+
+#[derive(Encode, Decode, Debug)]
+pub enum WorkerValidationError {
+	InternalError(String),
+	ValidationError(String),
+}
+
+#[derive(Encode, Decode, Debug)]
+pub enum ValidationResultHeader {
+	Ok(ValidationResult),
+	Error(WorkerValidationError),
+}
+
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+enum Mode {
+	Initialize,
+	Attach,
+}
+
+fn stringify_err(err: Box<dyn Error>) -> String {
+	format!("{:?}", err)
+}
+
+struct Inner {
+	shmem: Shmem,
+	candidate_ready_ev: Box<dyn EventImpl>,
+	result_ready_ev: Box<dyn EventImpl>,
+	worker_ready_ev: Box<dyn EventImpl>,
+
+	/// Flag that indicates that the worker side is attached to this workspace.
+	///
+	/// While there are apparent problems attaching multiple workers to the same workspace, we don't
+	/// need that anyway. So to make our reasoning a little bit simpler just add a flag and check
+	/// it before attaching.
+	attached: *mut AtomicBool,
+
+	/// The number of bytes reserved by the auxilary stuff like events from the beginning of the
+	/// shared memory area.
+	///
+	/// We expect this to be way smaller than the whole shmem size.
+	consumed: usize,
+}
+
+impl Inner {
+	fn layout(shmem: Shmem, mode: Mode) -> Self {
+		unsafe {
+			let base_ptr = shmem.as_ptr();
+			let mut consumed = 0;
+
+			let candidate_ready_ev = add_event(base_ptr, &mut consumed, mode);
+			let result_ready_ev = add_event(base_ptr, &mut consumed, mode);
+			let worker_ready_ev = add_event(base_ptr, &mut consumed, mode);
+
+			// The size of AtomicBool is guaranteed to be the same as the bool, however, docs
+			// on the bool primitve doesn't actually state that the in-memory size is equal to 1 byte.
+			//
+			// AtomicBool requires hardware support of 1 byte width of atomic operations though, so
+			// that should be fine.
+			//
+			// We still assert here to be safe than sorry.
+			static_assertions::assert_eq_size!(AtomicBool, u8);
+
+			// SAFETY: `AtomicBool` is represented by an u8 thus will be happy to take any alignment.
+			let attached = base_ptr.add(consumed) as *mut AtomicBool;
```

Comment:
Should `consumed` have 1 added to it in the case that the subsequent `align_up_to` is a no-op?

Reply:
Ah, good catch. We should always add 1.